### PR TITLE
feat(trace-view): Improve how we call eventstore.get_events

### PR DESCRIPTION
**Pull request by wmak**
- This allows eventstore.get_events to directly take a list of events
  to retrieve, currently get_events takes a filter of project and
  event ids.
  - This is problematic for the trace view where we already know the
    project to event_id mapping, and so we're not getting the
    optimization mentioned in the get_events function.

### CODE REVIEW

File: src/sentry/eventstore/snuba/backend.py [hunk 1]
```diff
@@ -92,26 +94,34 @@ def __get_events(
...
 
         # This is an optimization for the Group.filter_by_event_id query where we
         # have a single event ID and want to check all accessible projects for a
         # direct hit. In this case it's usually faster to go to nodestore first.
         if (
             filter.event_ids
             and filter.project_ids
-            and len(filter.event_ids) * len(filter.project_ids) < min(limit, NODESTORE_LIMIT)
+            and (
+                len(filter.event_ids) * len(filter.project_ids) < min(limit, NODESTORE_LIMIT)
```
Comment (wmak) [inline src/sentry/eventstore/snuba/backend.py:112]: so with trace-view, we can get up to 100 event_ids, and a large number of project_ids, but we'll ever only want to get up to 100 events, staying below this limit.
But since this is creating a list of events for every project + event we'll usually go over this limit and miss this optimization.

File: src/sentry/api/endpoints/organization_events_trace.py [hunk 2]
```diff
@@ -168,9 +169,14 @@ def serialize(self, parent_map, root, warning_extra, params, snuba_event=None, e
                 event.event_id: event
                 for event in eventstore.get_events(
                     eventstore.Filter(
+                        start=params["start"],
+                        end=params["end"],
```
Comment (wmak) [inline src/sentry/api/endpoints/organization_events_trace.py:173]: Unrelated fix, this corrects an issue where we're doing a 90d query in snuba for these events.

**Review by dashed — APPROVED** [review]

File: src/sentry/api/endpoints/organization_events_trace.py [hunk 3]
```diff
@@ -168,9 +169,14 @@ def serialize(self, parent_map, root, warning_extra, params, snuba_event=None, e
...
                         project_ids=params["project_id"],
-                        event_ids=[event["id"] for event in parent_map.values()],
-                    )
+                    ),
+                    event_list=[
+                        Event(project_id=event["project.id"], event_id=event["id"])
+                        for event in parent_map.values()
```
Comment (lynnagara) [inline src/sentry/api/endpoints/organization_events_trace.py:178]: The eventstore should create the actual event instances, like it does in the rest of this method. Doing it this way only if an event list is specified is rather inconsistent. I think it'd be better to just pass a list of project id / event id tuples.

File: src/sentry/eventstore/base.py [hunk 4]
```diff
@@ -146,6 +147,7 @@ def get_events(
         limit (int): Query limit - default 100
         offset (int): Query offset - default 0
         referrer (string): Referrer - default "eventstore.get_events"
+        event_list (Sequence[Event]): List of events to get
```
Comment (lynnagara) [inline src/sentry/eventstore/base.py:150]: project_id / event_id filtering is the role of the `snuba_filter`. Can we add this argument to the `snuba_filter` instead?

File: tests/sentry/eventstore/snuba/test_backend.py [hunk 5]
```diff
@@ -94,6 +95,27 @@ def test_get_events(self):
...
+            ),
+            event_list=[
+                Event(
+                    project_id=self.project1.id,
+                    event_id="a" * 32,
+                ),
+                Event(
+                    project_id=self.project2.id,
+                    event_id="b" * 32,
+                ),
```
Comment (markstory) [inline tests/sentry/eventstore/snuba/test_backend.py:111]: Seems curious to be giving a list of events to eventstore to get the events. Could there be another structure we use instead? Something like an `EventReference` or `EventIdentifier`?

Comment (wmak) [reply tests/sentry/eventstore/snuba/test_backend.py:111]: Hmm good point :+1: I'll create `EventIdentifier`, I think get_events can return stubbed events which might get confused with `EventReference`

File: src/sentry/eventstore/models.py [hunk 6]
```diff
@@ -29,6 +29,21 @@ def ref_func(x):
...
+    """
+
+    __slots__ = ["project_id", "event_id"]
+
+    def __init__(self, project_id, event_id):
+        self.project_id = project_id
+        self.event_id = event_id
+
+    def to_event(self):
+        return Event(self.project_id, self.event_id)
```
Comment (lynnagara) [inline src/sentry/eventstore/models.py:44]: Please don't introduce a dependency of the Snuba filter to the Snuba event model (filter shouldn't know anything about the model). Eventstore should be the one creating the Event objects like it for all of the rest of the methods. I'd suggest making EventIdentifier a tuple (or a named tuple if you prefer) and move it out of the models.py file.

**Review by lynnagara — CHANGES_REQUESTED** [review]

**Comment by wmak** [comment]
Closing this in favour of #24160, see PR description there for explanation of why
